### ISSUE TITLE
feat: Switch to using managed Iterators

### ISSUE THREAD

**Issue by adityapk00** [farcasterxyz/hub-monorepo#1122]
**What is the feature you would like to implement?**
Implement iterators as callbacks, so that the iterator is always managed by the DB class, so that we can always ensure that it is closed

**Why is this feature important?**
Our usage of iterators today leaves the possibility that an iterator may never be closed, which leads to a lot of downstream DB issues. Additionally, we need to code iterators very carefully, which is prone to bugs. 

**Will the protocol spec need to be updated??**
No protocol impact

**How should this feature be built? (optional)**
```
db.forEachIterator(prefix?, ([key, value]) => {
 // do stuff with key value. 
 // 1. if this method throws, we close the iterator
 // 2. If all [key, values] are processed, we close the iterator
 // 3. If the (optional) timeout is hit, we close the iterator
}, timeout?);
```

**Notes**
1. timeout default should be 1 min

**Comment by varunsrin** [comment]
@adityapk00 fixed several leaky iterators [here](https://github.com/farcasterxyz/hub-monorepo/pull/1134) which shipped in v1.4.1, but there are still more fixes needed.

**Comment by adityapk00** [comment]
https://github.com/farcasterxyz/hub-monorepo/pull/1156
